x/valset-pref/validator_set.go

```diff
@@ -170,53 +239,64 @@ func (k Keeper) PreformRedelegation(ctx sdk.Context, delegator sdk.AccAddress, e
 	}
 
 	// calculate the difference between two sets
-	var diffValSet []*valSet
+	var diffValSets []*valSet
 	for i, newVals := range existingValSet {
-		diffAmount := newVals.amount.Sub(newValSet[i].amount)
+		diffAmount := newVals.Amount.Sub(newValSet[i].Amount)
 
 		diff_val := valSet{
-			valAddr: newVals.valAddr,
-			amount:  diffAmount,
+			ValAddr: newVals.ValAddr,
+			Amount:  diffAmount,
 		}
-		diffValSet = append(diffValSet, &diff_val)
-	}
-
-	// Algorithm starts here
-	for _, diff_val := range diffValSet {
-		for diff_val.amount.TruncateDec().GT(sdk.NewDec(0)) {
-			source_val := diff_val.valAddr
-			target_val, idx := k.FindMin(diffValSet, source_val)
-
-			// checks if there are any more redelegation possible
-			if target_val.amount.TruncateDec().Equal(sdk.NewDec(0)) {
-				break
-			}
-
-			validator_source, err := sdk.ValAddressFromBech32(source_val)
-			if err != nil {
-				return fmt.Errorf("source validator address not formatted")
+		diffValSets = append(diffValSets, &diff_val)
+	}
+
+	// Algorithm starts here, verbose explanation in README.md
```

Comment:
Algorithm has been tested in depth in msg_server_test.go, that's why i didn't add a keeper test for it